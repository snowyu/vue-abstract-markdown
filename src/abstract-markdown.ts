import { Vue, Component, Prop, Watch } from 'vue-property-decorator';
import { CreateElement } from 'vue';

/**
 * AbstractMarkdown Viewer Mixin
 * @abstract
 * @property source: the markdown source.
 * @property visible: whether render the markdown. defaults to true
 * @property highlight: whether enable syntax highlight. defaults to true
 * @property html: whether enable html. defaults to true
 * @property xhtmlOut: whether enable xhtml. defaults to true
 * @property breaks: whether convert '\n' to '<br>'. defaults to true.
 * @property linkify: whether autoconvert URL-like text to links. defaults to true
 * @property emoji: whether convert emotion symbol to '<img emotion>', defaults to true
 * @property typographer: whether enable some language-neutral replacement + quotes beautification, defaults to true
 * @property quotes: Double + single quotes replacement pairs, when typographer enabled, defaults to '“”‘’'
 *
 */
@Component({
  components: {
  }
})
export class AbstractMarkdown extends Vue {
  @Prop({type: String, default: ''}) source!: string;
  @Prop({type: Boolean, default: true}) visible!: boolean;
  @Prop({type: Boolean, default: true}) highlight!: boolean;
  @Prop({type: Boolean, default: true}) html!: boolean;
  @Prop({type: Boolean, default: true}) xhtmlOut!: boolean;
  @Prop({type: Boolean, default: true}) breaks!: boolean;
  @Prop({type: Boolean, default: true}) linkify!: boolean;
  @Prop({type: Boolean, default: true}) emoji!: boolean;
  @Prop({type: Boolean, default: true}) typographer!: boolean;
  @Prop({type: String, default: '“”‘’'}) quotes!: string;
  sourceData = this.source;

  @Watch('source') @Watch('visible')
  protected _onBeforeRender() {
    this.sourceData = this.prerender();
    this.$forceUpdate()
  }

  prerender(content = this.source) {
    /* istanbul ignore next */
    if (this.emoji) {
      content = this.parseEmoji(content)
    }
    return content;
  }

  toHtml(text: string, options?){
    return text;
  }

  parseEmoji(text: string) {
    return text;
  }

  render(h: CreateElement) {
    const vContent = this.visible ? this.toHtml(this.sourceData, this) : '';
    return h('div', {
      domProps: {
        innerHTML: vContent,
      },
    })
  }

  beforeMount() {
    /* istanbul ignore next */
    if (this.$slots.default) {
      let vSource = '';
      for (const slot of this.$slots.default) {
        vSource += slot.text;
      }
      if (vSource) {
        this.sourceData = this.prerender(vSource);
      }
    }
  }
}
