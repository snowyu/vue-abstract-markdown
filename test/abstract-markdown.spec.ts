/**
 * @jest-environment jsdom
 */

jest.mock('../src/environment.ts', () => ({
  IS_DEV: true,
  IS_PROD: false,
}))

import { AbstractMarkdown } from '../src/abstract-markdown'
import Vue, { VueConstructor } from "vue/dist/vue";
import { Mixins, Component } from 'vue-property-decorator';
import { mount } from '@vue/test-utils';

describe(`AbstractMarkdown`, () => {
  @Component
  class Md extends AbstractMarkdown {
    toHtml(text: string, options?){
      return text.replace(/\*\*/g, '--');
    }
    parseEmoji(text: string) {
      return text.replace(/\:\)/g, '<img smile>');
    }
  }

  beforeEach(() => {
    // markdown = Vue.extend(AbstractMarkdown);
  })

  it(`should create abstract markdown component`, () => {
    const vOpt = {
      components: { AbstractMarkdown },
      template: '<abstract-markdown>hello **world**</abstract-markdown>'
    };
    const Comp = Vue.extend({
      components: { AbstractMarkdown },
      template: '<abstract-markdown>hello **world**</abstract-markdown>'
    });
    // const wrapper = mount(vOpt);
    const vComp = new Comp().$mount();
    expect(vComp.$el).toMatchSnapshot();
  })

  it(`should mixin the abstract markdown component with slot`,  () => {
    const vOpts = {
      components: { Md },
      template: '<md>hello:) **world**</md>'
    };
    const wrapper = mount(vOpts);
    const vm = wrapper.vm;
    expect(vm.$el).toMatchSnapshot();
  })

  it(`should mixin the abstract markdown component with source prop`,  (done) => {
    const vOpts = {
      components: { Md },
      data: function() {
        return {
          text: 'hello pro **world**'
        }
      },
      template: '<md :source="text"></md>'
    };
    const wrapper = mount(vOpts);
    const vm = wrapper.vm;

    expect(vm.$el).toMatchSnapshot();
    vm['text'] = 'ok it **as**';
    vm.$nextTick(()=>{
      expect(vm.$el).toMatchSnapshot();
      done();
    })
  })

  it(`should mixin the abstract markdown component with invisible`,  () => {
    const vOpts = {
      components: { Md },
      data: function() {
        return {
          visible: true
        }
      },
      template: '<md :visible="visible">hello **world**</md>'
    };
    const wrapper = mount(vOpts);
    const vm = wrapper.vm;
    vm['visible'] = false;
    expect(vm.$el).toMatchSnapshot();
  })

})
