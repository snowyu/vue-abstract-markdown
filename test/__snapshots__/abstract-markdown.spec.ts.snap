// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbstractMarkdown should create abstract markdown component 1`] = `
<div>
  hello **world**
</div>
`;

exports[`AbstractMarkdown should mixin the abstract markdown component with invisible 1`] = `<div />`;

exports[`AbstractMarkdown should mixin the abstract markdown component with slot 1`] = `
<div>
  hello
  <img
    smile=""
  />
   --world--
</div>
`;

exports[`AbstractMarkdown should mixin the abstract markdown component with source prop 1`] = `
<div>
  hello pro --world--
</div>
`;

exports[`AbstractMarkdown should mixin the abstract markdown component with source prop 2`] = `
<div>
  ok it --as--
</div>
`;
